File: src/services/inbox.service.ts
import { roundRepository } from '../repositories/round.repository.js';
import { chitRepository } from '../repositories/chit.repository.js';
import { aliasService } from './alias.service.js';
import { AppError } from '../errors/app-error.js';
import { ERROR_CODES } from '../errors/error-codes.js';
import { AnonymousInboxDTO } from '../dto/inbox.dto.js';
import { RoundStatus } from '../types/enums.js';

export const inboxService = {
  async getAnonymousInbox(roomId: string, roundId: string): Promise<AnonymousInboxDTO> {
    const round = await roundRepository.findById(roundId);
    if (!round) throw AppError.notFound(ERROR_CODES.ROUND_NOT_FOUND, 'Round not found');

    if (round.status === RoundStatus.waiting) {
      throw AppError.forbidden(ERROR_CODES.SUBMISSIONS_NOT_OPEN, 'Round has not started yet');
    }

    const chits = await chitRepository.findByRoundId(roundId);
    const mappings = await aliasService.getAliasMapping(roundId, round.aliasEpoch);

    const anonymousChits = chits.map(chit => {
      const alias = mappings[chit.senderId];
      if (!alias) return null;

      return {
        anonymousChitId: alias.opaqueAliasId + '-' + chit.id, // Compound ID
        alias: `Player ${alias.aliasNumber}`,
        body: chit.body,
        submittedAt: chit.submittedAt,
        isRead: chit.isRead,
        isGuessed: chit.isGuessed,
      };
    }).filter(c => c !== null);

    return {
      roundNumber: round.roundNumber,
      aliasEpoch: round.aliasEpoch,
      chits: anonymousChits as any,
    };
  },

  async markChitRead(anonymousChitId: string, roomId: string, isRead: boolean): Promise<void> {
    const chitId = anonymousChitId.split('-').slice(1).join('-');
    if (!chitId) throw AppError.badRequest(ERROR_CODES.CHIT_NOT_FOUND, 'Invalid chit ID');

    const chit = await chitRepository.update(chitId, { body: undefined } as any);
    if (!chit) throw AppError.notFound(ERROR_CODES.CHIT_NOT_FOUND, 'Chit not found');

    await chitRepository.markRead(chitId, isRead);
  },

  async markChitGuessed(anonymousChitId: string, roomId: string, isGuessed: boolean, guessedAliasId?: string): Promise<void> {
    const chitId = anonymousChitId.split('-').slice(1).join('-');
    if (!chitId) throw AppError.badRequest(ERROR_CODES.CHIT_NOT_FOUND, 'Invalid chit ID');

    let participantId: string | undefined;
    if (isGuessed && guessedAliasId) {
      const resolved = await aliasService.resolveAliasToParticipant(guessedAliasId);
      if (resolved) participantId = resolved;
    }

    await chitRepository.markGuessed(chitId, isGuessed, participantId);
  }
};

File: app/services/inbox_service.py
from typing import Optional
from sqlalchemy.ext.asyncio import AsyncSession
from sqlalchemy import select

from app.core.exceptions import NotFoundException
from app.models.chit import ChitMessage
from app.models.round import GameRound
from app.schemas.inbox import (
    AnonymousChitDTO,
    AnonymousInboxDTO,
    MarkChitGuessedDTO,
    MarkChitReadDTO,
)
from app.services.alias_service import (
    get_or_create_aliases_for_epoch,
    resolve_opaque_alias_id,
)

async def get_anonymous_inbox(db: AsyncSession, room_id: str) -> AnonymousInboxDTO:
    stmt_round = (
        select(GameRound)
        .where(GameRound.roomId == room_id)
        .order_by(GameRound.roundNumber.desc())
    )
    res_round = await db.execute(stmt_round)
    current_round = res_round.scalars().first()

    if not current_round:
        return AnonymousInboxDTO(roundNumber=0, aliasEpoch=0, chits=[])

    aliases = await get_or_create_aliases_for_epoch(db, current_round)
    alias_map = {a.participantId: a for a in aliases}

    stmt_chits = select(ChitMessage).where(ChitMessage.roundId == current_round.id)
    res_chits = await db.execute(stmt_chits)
    chits = list(res_chits.scalars().all())

    dto_list = []
    for chit in chits:
        alias = alias_map.get(chit.senderId)
        alias_number = alias.aliasNumber if alias else 0
        opaque_id = alias.opaqueAliasId if alias else chit.id

        dto_list.append(
            AnonymousChitDTO(
                anonymousChitId=opaque_id,
                alias=f"Player {alias_number}",
                body=chit.body,
                submittedAt=chit.submittedAt,
                isRead=chit.isRead,
                isGuessed=chit.isGuessed,
            )
        )

    dto_list.sort(key=lambda x: int(x.alias.split(" ")[1]) if " " in x.alias else 0)

    return AnonymousInboxDTO(
        roundNumber=current_round.roundNumber,
        aliasEpoch=current_round.aliasEpoch,
        chits=dto_list,
    )

async def mark_chit_read(
    db: AsyncSession,
    room_id: str,
    opaque_alias_id: str,
    is_read: bool,
) -> MarkChitReadDTO:
    alias = await resolve_opaque_alias_id(db, opaque_alias_id)
    sender_id = alias.participantId if alias else opaque_alias_id

    stmt_round = (
        select(GameRound)
        .where(GameRound.roomId == room_id)
        .order_by(GameRound.roundNumber.desc())
    )
    res_round = await db.execute(stmt_round)
    current_round = res_round.scalars().first()

    if not current_round:
        raise NotFoundException("Round not found.")

    stmt_chit = select(ChitMessage).where(
        ChitMessage.roundId == current_round.id,
        ChitMessage.senderId == sender_id,
    )
    res_chit = await db.execute(stmt_chit)
    chit = res_chit.scalars().first()

    if chit:
        chit.isRead = is_read
        db.add(chit)
        await db.flush()

    return MarkChitReadDTO(isRead=is_read)

async def mark_chit_guessed(
    db: AsyncSession,
    room_id: str,
    opaque_alias_id: str,
    is_guessed: bool,
    guessed_alias_id: Optional[str] = None,
) -> MarkChitGuessedDTO:
    alias = await resolve_opaque_alias_id(db, opaque_alias_id)
    sender_id = alias.participantId if alias else opaque_alias_id

    stmt_round = (
        select(GameRound)
        .where(GameRound.roomId == room_id)
        .order_by(GameRound.roundNumber.desc())
    )
    res_round = await db.execute(stmt_round)
    current_round = res_round.scalars().first()

    if not current_round:
        raise NotFoundException("Round not found.")

    stmt_chit = select(ChitMessage).where(
        ChitMessage.roundId == current_round.id,
        ChitMessage.senderId == sender_id,
    )
    res_chit = await db.execute(stmt_chit)
    chit = res_chit.scalars().first()

    if chit:
        chit.isGuessed = is_guessed
        if guessed_alias_id:
            target_alias = await resolve_opaque_alias_id(db, guessed_alias_id)
            chit.guessedParticipantId = target_alias.participantId if target_alias else guessed_alias_id
        db.add(chit)
        await db.flush()

    return MarkChitGuessedDTO(isGuessed=is_guessed)


